internal/handler_swap.go

```diff
@@ -18,23 +18,101 @@ package internal
 
 import (
 	"bufio"
+	"errors"
 	"fmt"
-	"os"
+	"log"
 	"os/exec"
 	"strconv"
 	"strings"
+
+	"github.com/spf13/afero"
 )
 
+// logger is an interface to substitute *log.Logger for unit-tests.
+type logger interface {
+	Println(v ...any)
+}
+
+var _ logger = &log.Logger{}
+
+// execCommander is an interface to substitute exec.Command() for unit-tests.
+type execCommander interface {
+	ExecAndOutput(name string, arg ...string) ([]byte, error)
+}
+
+var _ execCommander = realExecCommander{}
+
+type realExecCommander struct{}
+
+func (c realExecCommander) ExecAndOutput(name string, arg ...string) ([]byte, error) {
+	return exec.Command(name, arg...).Output()
+}
+
+// NewSwap is a constructor for Swap.
+func NewSwap(
+	defaultSwapSizeBytes int64,
+	availableSwapSizes []string,
+	oldSwappinessUnitFile string,
+	defaultSwapFileLocation string,
+	fs afero.Fs,
+	loggerInfo logger,
+) (*Swap, error) {
+	if defaultSwapSizeBytes == 0 {
+		return nil, errors.New("defaultSwapSizeBytes is required")
+	}
+	if len(availableSwapSizes) == 0 {
+		return nil, errors.New("availableSwapSizes is required")
+	}
+	if oldSwappinessUnitFile == "" {
+		return nil, errors.New("oldSwappinessUnitFile is required")
+	}
+	if loggerInfo == nil {
+		return nil, errors.New("info logger is required")
+	}
+	if defaultSwapFileLocation == "" {
+		return nil, errors.New("default swap location is required")
+	}
+	if fs == nil {
+		return nil, errors.New("fs is required")
+	}
+	swapFileLocation, err := getSwapFileLocation(fs, defaultSwapFileLocation)
+	if err != nil {
+		return nil, fmt.Errorf("getting swapfile location: %w", err)
+	}
+
+	return &Swap{
+		defaultSwapSizeBytes:  defaultSwapSizeBytes,
+		availableSwapSizes:    availableSwapSizes,
+		oldSwappinessUnitFile: oldSwappinessUnitFile,
+		swapFileLocation:      swapFileLocation,
+		fs:                    fs,
+		execCommander:         realExecCommander{},
+		loggerInfo:            loggerInfo,
+	}, nil
+}
+
+// Swap decorates all functionality related to swap/swappiness changes.
+type Swap struct {
+	defaultSwapSizeBytes  int64
+	availableSwapSizes    []string
+	oldSwappinessUnitFile string
+	swapFileLocation      string
+	fs                    afero.Fs
+	execCommander         execCommander
+	loggerInfo            logger
+}
+
 // Get swap file location from the system (/proc/swaps)
 // Sample output:
 // Filename				Type		Size	Used	Priority
 // /home/swapfile			file		8388604	0	-2
-func getSwapFileLocation() (string, error) {
-	file, err := os.Open("/proc/swaps")
+func getSwapFileLocation(fs afero.Fs, defaultSwapFileLocation string) (string, error) {
+	filepath := "/proc/swaps"
+	file, err := fs.Open(filepath)
 	if err != nil {
 		return "", err
```

Comment:
I've also noticed that this scenario does not fallback to a default swap file location.
Not sure if it was done intentionally.
It is easy to change with a `defer` if needed